models/user_openid.go

```diff
@@ -0,0 +1,152 @@
+// Copyright 2017 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package models
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"code.gitea.io/gitea/modules/auth/openid"
+)
+
+var (
+	// ErrOpenIDNotExist openid is not known
+	ErrOpenIDNotExist = errors.New("OpenID is unknown")
+)
+
+// UserOpenID is the list of all OpenID identities of a user.
+type UserOpenID struct {
+	ID          int64  `xorm:"pk autoincr"`
+	UID         int64  `xorm:"INDEX NOT NULL"`
+	URI         string `xorm:"UNIQUE NOT NULL"`
+}
+
+// GetUserOpenIDs returns all openid addresses that belongs to given user.
+func GetUserOpenIDs(uid int64) ([]*UserOpenID, error) {
+	openids := make([]*UserOpenID, 0, 5)
+	if err := x.
+		Where("uid=?", uid).
+		Find(&openids); err != nil {
+		return nil, err
+	}
+
+	return openids, nil
+}
+
+func isOpenIDUsed(e Engine, uri string) (bool, error) {
+	if len(uri) == 0 {
+		return true, nil
+	}
+
+	return e.Get(&UserOpenID{URI: uri})
+}
+
+// IsOpenIDUsed returns true if the openid has been used.
+func IsOpenIDUsed(openid string) (bool, error) {
+	return isOpenIDUsed(x, openid)
+}
+
+func addUserOpenID(e Engine, openid *UserOpenID) error {
+	openid.URI = strings.ToLower(strings.TrimSpace(openid.URI))
+	used, err := isOpenIDUsed(e, openid.URI)
+	if err != nil {
+		return err
+	} else if used {
+		return ErrOpenIDAlreadyUsed{openid.URI}
+	}
+
+	_, err = e.Insert(openid)
+	return err
+}
+
+// AddUserOpenID adds an openid address to given user.
+func AddUserOpenID(openid *UserOpenID) error {
+	return addUserOpenID(x, openid)
+}
+
+// AddUserOpenIDs adds an openid address to given user.
+func AddUserOpenIDs(openids []*UserOpenID) error {
+	if len(openids) == 0 {
+		return nil
+	}
+
+	// Check if any of them has been used
+	for i := range openids {
+		// TODO: normalize ?
+		openids[i].URI = strings.ToLower(strings.TrimSpace(openids[i].URI))
+		used, err := IsOpenIDUsed(openids[i].URI)
```

Comment:
setup a `sess := x.NewSession()`, `defer sess.Close()` outside the loop, and use `isOpenIDUsed(sess, openids[i].URI)`
Otherwise we end up with race-conditions 💥